src/dataflow/stream/internal_stateful_read_stream.rs

```diff
@@ -76,17 +80,19 @@ impl<D: Data, S: State> EventMakerT for InternalStatefulReadStream<D, S> {
                 // Stateful callbacks may not run in parallel because they access shared state,
                 // so create 1 callback for all
                 let stateful_cbs = self.callbacks.clone();
-                let mut state_ref = Arc::clone(&self.state);
+                let mut state_arc = Arc::clone(&self.state);
                 if !stateful_cbs.is_empty() {
                     events.push(OperatorEvent::new(
                         msg.timestamp.clone(),
                         false,
                         move || {
                             for callback in stateful_cbs {
                                 let msg_copy = msg.clone();
-                                (callback)(msg_copy.timestamp, msg_copy.data, unsafe {
-                                    Arc::get_mut_unchecked(&mut state_ref)
-                                })
+                                let state_ref_mut =
+                                    unsafe { Arc::get_mut_unchecked(&mut state_arc) };
+                                state_ref_mut.set_access_context(AccessContext::Callback);
+                                state_ref_mut.set_current_time(msg.timestamp.clone());
+                                (callback)(msg_copy.timestamp, msg_copy.data, state_ref_mut)
```

Comment:
Is this safe in a multi-threaded operator? Since multiple mutable references to `self.state` are allowed and there is no locking in the state data structure, is it possible that we set the AccessContext as Callback and start a message callback for time `t + 1`, meanwhile, also start a watermark callback for time `t`, thus changing the AccessContext to WatermarkCallback and allow access to state in the message callback? The code in the operator to achieve this might be a bit contrived, but I think it's possible to do this.

Reply:
Similarly, the `current_time` might jump around in a multi-threaded operator.

Reply:
Hmm, I thought that for now only 1 callback can access state at a time. I agree about that scenario for the future --  a better solution might be to clone (could be expensive) or provide a view on the datastructure.

Reply:
Can you please open an issue to track this?